### ISSUE TITLE
Crafting and pickit

### ISSUE THREAD

**Issue by thatguy692** [kolton/d2bot-with-kolbot#749]
I'm confused on what to do in order for the bot to collect the ingredients to craft blood gloves, so far i've:
-enabled cubing
-enabled blood gloves recipe
My question is: will the bot automatically start collecting the necessary ingredients: jewel, rune and make the p gem?
will it also collect the base item needed for the craft or do i need to specifically add it to my pickit file?

**Comment by DarkHorseDre** [comment]
I've not crafted, but notes in config state `// Ingredients for the following recipes will be auto-picked, for classids check libs/NTItemAlias.dbl`

as for gems - I cant check the scripts now so cant tell if it would cube flawless to pgems or only pick pgems. If you don't check the code and don't get a solid answer here, test it! or have it cube flawless to perfect just incase..

**Comment by thatguy692** [comment]
the thing is the bot is picking up flawless gems and cubing them to perfect but it didn't pick any runes or jewels or base item for craft so i don't really know

**Comment by DarkHorseDre** [comment]
check cubing.js
`case Recipe.Blood.Gloves:
				this.recipes.push({Ingredients: [Config.Recipes[i][1], 613, 643, 581], Level: 79, Index: Recipe.Blood.Gloves});`

looks like it wants level 79+ jool - what difficulty and area are you farming?

**Comment by DarkHorseDre** [comment]
please post a pastebin link to your config

I'd try and run only one cubing recipe for now as I had issues with multi recently (so no charm rolling, no socketing eth weps/armour etc)

ensure you have space in stash obvs!

**Comment by thatguy692** [comment]
i have charm rolling enabled. here's my config:
https://pastebin.com/tee4raRL

**Comment by DarkHorseDre** [comment]
so you gonna try narrow it down like i suggested?

also note you can check the console, as per:
    Config.ShowCubingInfo = true; // show/hide the cubing messages on console and item log screen

**Comment by thatguy692** [comment]
i disabled charm rolling so theres only 1 recipe active now

**Comment by DaCapoAlCoda** [comment]
I've not use the crafting option yet, but when cubing runes, I have to have runes and gems set in pickit, though the bot will automatically cube when needed recipe is collected.

**Comment by DarkHorseDre** [comment]
@DaCapoAlCoda pickit tells bot what to keep.
crafting will collect ingredients even if not in pickit to make target item (see the comments in char config which I pasted in post above)

if target item is not in pickit, bot will make target item then drop it.

eg. for fire/cold life skiller, you need `[name] == grandcharm && [quality] == magic # [fireskilltab]+[coldskilltab] == 1` in pickit.  if not, crafting will make it then drop it.

You dont need to specify anything else, just this line in config (in addition to enabling `Config.Cubing = true;`):
`Config.Recipes.push([Recipe.Reroll.Magic, "Grand Charm"]);`

**Comment by ughmas** [comment]
I am experiencing a similar issue when trying to craft Caster boots, however my Caster Amulet and Blood Ring recipies work just fine... Do we need to add the base Item Alias in place of the Config.Recipes[i][1] in cubing.js or is this automatically specifying the correct base item? (example 455 for wyrmhideboots) 

I have these three recipies enabled, it picks up rals and sols, jewels and ammys, amythests and rubys, crafts the rings and ammys fine - but does not pick Thuls or boots/demonhide boots/wyrmhide boots.

Also just to re-iterate above, you do not need to have the items in your pickit file (runes etc) they are auto picked- at least for the ones that are working.

**Comment by DarkHorseDre** [comment]
@ughmas yeah i think you're proving an issue as it is collecting correct base and ingredients for caster amu but not crafter boots (assuming you've not customised any of the files).

This is from my cubing.js - I traced through the ingredients (619, 643, 561) to check. The boots you are finding need to be level 71+ as per the below line:

`this.recipes.push({Ingredients: [Config.Recipes[i][1], 619, 643, 561], Level: 71, Index: Recipe.Caster.Boots});`

but seems the boots are missing, look at caster amu and ring recipe - 4 items, not 3:
`case Recipe.Caster.Amulet:
`		this.recipes.push({Ingredients: [**520**, 617, 643, 561], Level: 90, Index: Recipe.Caster.Amulet});`

`case Recipe.Caster.Ring:
		this.recipes.push({Ingredients: [**522**, 620, 643, 561], Level: 77, Index: Recipe.Caster.Ring});`

From NTItemAlias.dbl:
619 =  thul
643 = jewel
561 = perfectamethyst
[no boots here brah]

522 = ring
520 = amulet

455 = wyrmhide boots

I think adding the boots as the first parameter (of 4) should work

**Comment by DarkHorseDre** [comment]
@thatguy692 - If I'm right in what I've found, you're missing the base ingredient too as theres no base mentioned for your blood gloves - try adding it

**Comment by DarkHorseDre** [comment]
yeah ignore the modified comment - I was first gonna tell you to go look but then ended up looking myself :)

**Comment by ughmas** [comment]
I guess one more thing to ponder while I am testing is, what exactly is "Config.Recipes[i][1]" referencing? and if you do need to replace this with the item alias - how would you add multiple items for the base if this works? i.e/ you wanted to craft both demonhide boots AND wyrmhide - maybe just put them in parenthesis (385, 455)?
not sure - this is the only part of the coding that is confusing to me

**Comment by DarkHorseDre** [comment]
hmmm thats a good question. im now starting to think that the items are specified elsewhere - seemed too much of an obvious omission as others are missing too (where the required base can vary. check armours too: http://classic.battle.net/diablo2exp/items/crafted/caster.shtml)

Are you sure you are farming in area level 71+ to drop these bases?

I cant find evidence of where-else these items are specified.  unless someone who knows comments here, I would do the quickest thing and hard code 455 and if that works, troubleshoot.

My feeling is that the bot knows which bases to collect for the recipe but something else is stopping it from collecting the boots

**Comment by adamveg** [comment]
> 
> 
> hmmm thats a good question. im now starting to think that the items are specified elsewhere - seemed too much of an obvious omission as others are missing too (where the required base can vary. check armours too: http://classic.battle.net/diablo2exp/items/crafted/caster.shtml)
> 
> Are you sure you are farming in area level 71+ to drop these bases?
> 
> I cant find evidence of where-else these items are specified. unless someone who knows comments here, I would do the quickest thing and hard code 455 and if that works, troubleshoot.
> 
> My feeling is that the bot knows which bases to collect for the recipe but something else is stopping it from collecting the boots

Did you find out where codes for crafting are ? :)

**Comment by jogohnwer** [comment]
their in C:\Users\JD\Desktop\d2bot-with-kolbot\d2bs\kolbot\libs\NTItemAlias.dbl on my computer
